### ISSUE TITLE
Design a better grouping system for Utterance

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/scenery-phet#485]
Alert "grouping" is currently the way that we prevent an overwhelming amount of alerts to be sent to assistive devices.

In discussion with @samreid and @zepumph we identified a few shortcomings and ways that this system could be improved.

First, we dislike how delayTime and uniqueGroupId work. uniqueGroupId is the way that an alert group is specified. If more than one alert of the same group is in the queue, the older alert is removed so that the user receives the most recent information for that alert category. `delayTime` makes it so that an alert is in the queue for a certain amount of time before before being spoken. When these two are used in combination, the delayed utterance will be removed immediately and the utterance of a group won't be spoken until new utterances stop getting added to the queue.  This is very confusing and feels like a side effect. This feature (or something like it should be better designed into Utterance.

One improvement could be that some utterances have a timer interval specified such that when utterances hit the queue rapidly we wait to send the alert content to the AT until changes have stopped. As a side note, this may mean that the `uniqueGroup` concept should be reconsidered entirely for this system instead.

Another improvement that was discussed would be to automatically prevent duplicate alerts with the same content. We discussed potentially using a map to make this very fast. But for now we likely don't have very many utterances in the queue, so we can probably get away with just looping through the array of utterances.

Finally, we discussed tracking utterances that had been previously spoken and preventing alerts with the same content from being announced. But we decided this is something we cannot do because it will be impossible to determine when we actually want this. This could prevent the most up to date alert information from being spoken to the user.

**Comment by jessegreenberg** [comment]
> One improvement could be that some utterances have a timer interval specified such that when utterances hit the queue rapidly we wait to send the alert content to the AT until changes have stopped

Here is a proposal:
We remove both options `uniqueGroupId` and `delayTime`. They will become implementation details. `Utterance` has already been implemented so that it a client can keep a reference to it and add it to the `utteranceQueue` multiple times. We add a new boolean option called `alertWhenStable` (or something). When true the utterance will receive a unique ID and a delay of some amount of time. If a new utterance is added with the same unique id before the old one is removed from the queue, we remove the old alert and the new delay begins to wait for the alert to be "stable". The amount of time could be changeable, but that may not be necessary.

I guess the point is this works exactly as the combination of `delayTime` and `uniqueGroupId` currently do together, but the API is far nicer.

@zepumph what do you think? Is this all we need for that part?

**Comment by zepumph** [comment]
> but the API is far nicer.

You can say that again! I really like this. And by making this an `Utterance`-centric implementation, we can easily supply options/add options when we create the utterance. Before going through with this I looked at the `new Utterance( { alert: alert, uniqueGroupId: 'position' } )`, which is used hierarchically. I'm sure this can be factored out, but it is good to note the only case I could find in the project where the same `uniqueGroupId` was used across files.

With that in mind, and perhaps a better name than `alertWhenStable`, we are all steams ahead.

**Comment by jessegreenberg** [comment]
Hmm, `collapse`, `collapseAlerts`, `uniqueAlert`, `oneOffAlert`, `loneAlert`, `singleAlert`, `singularAlert`, `theChosenOne`

**Comment by jessegreenberg** [comment]
> only case I could find in the project where the same uniqueGroupId was used across files.

Good point, I hadn't through of that. I guess the way it would work is that the reference to that `Utterance` would be shared, or stored on the supertype.

**Comment by jessegreenberg** [comment]
More names:

>When this is true, alerts will wait to be spoken until this utterance stops being added
to the queue for X ms. This will the same utterance from hitting the user too frequently.

alertFixed
alertStatic
alertStable
preventSpam
temperAlerts
suppressRepeats

**Comment by jessegreenberg** [comment]
This was mentioned during disccussion, but this proposal  has the limitation that if the alert *never* stops changing, it will *never* be spoken.

**Comment by jessegreenberg** [comment]
Maybe we should go a bit further with changes than I proposed in https://github.com/phetsims/scenery-phet/issues/485#issuecomment-481057247. More brainstorming:
```js
class UtteranceSupressor {
  constructor( options ) {

    options = _.extend( {

      // {boolean} if true, the alert will not be spoken until this utterance stops being added to the
      // utteranceQueue for stableDelay time, or if alertMinimumFrequency time has passed while this 
      // utterance has continuously changed (if it is specified).
      alertStable: true,

      // {number} - in ms, how long to wait before the utterance is considered "stable" and it is announced
      alertStableDelay: 500,

      // {null|number} - if specified, the utterance will be spoken at least this frequently in ms
      // even if the utterance is continuously added to the queue and never becomes stable
      alertMinimumFrequency: null,

      // @returns {boolean} - if predicate returns false, the alert content associated
      // with the utterance will not be announced
      predicate: function() { return true; },
    }, options );

    if ( assert && options.alertMinimumFrequency ) {
      assert( options.alertStableDelay < alertMinimumFrequency );
    }

    // ...
  }
}
```

An UtteranceSuppressor can be passed through options and by default an UtteranceSuppressor with its default options is used.

**Comment by zepumph** [comment]
To me all of these options could be added to an `Utterance` just fine. I don't see a sub-type as that beneficial. I'm a bit confused about the predicate option. Could you provide a bit more explanation including a potential use case?

**Comment by jessegreenberg** [comment]
OK, sounds good. My motivation for creating a type was to make it very explicit that these options are dependent on one another.

The `predicate` function is already a feature of Utterance, but I thought it would fit into a "Suppressor" type if we had one. The purpose it to provide another way to prevent stale alerts. So you could do something like

```js
predicate: () => {
  // only announce this utterance the object still exists in the play area
  return playArea.contains( myThing );
}
```

**Comment by zepumph** [comment]
To me, I think of an api similar to `minify.js`. Not to say this is the "best" way to do it, but here "parallel leveled" options are dependent on each other and I think it makes sense, and is easy to document:

```js
const MINIFY_DEFAULTS = {
  minify: true,

  // Only enabled if minify is true
  babelTranspile: true,
  uglify: true,

  // Only enabled if uglify is true
  mangle: true,
  stripAssertions: true,
  stripLogging: true,
  beautify: false
};

```

And so I was thinking of something similar for `Utterance`.  We would have some options "only enabled if `alertStable` is true."

>I thought some of these things might not be optional

To me each of these could have a nice default. If you disagree please explain.

**Comment by jessegreenberg** [comment]
That sounds great to me, I will add them as options in Utterance.

**Comment by jessegreenberg** [comment]
Now that Utterances are designed to be persistent and their `alert` can be changed, I am finding that the required alert is too strict. It is more convenient to set the alert later. @zepumph is it OK if I remove
```js
      assert && assert( config.alert, 'alert is required' );
```

**Comment by jessegreenberg** [comment]
I am going to proceed with that, which means that config can be turned to options since none of the fields are required.

**Comment by jessegreenberg** [comment]
From the original  ticket:
> One improvement could be that some utterances have a timer interval specified such that when utterances hit the queue rapidly we wait to send the alert content to the AT until changes have stopped

I believe this is done. The other potential improvement for this issue was

> One improvement could be that some utterances have a timer interval specified such that when utterances hit the queue rapidly we wait to send the alert content to the AT until changes have stopped

I don't think this is necessary at this time. A similar outcome can be accomplished now by using `alertStable` for an Utterance.  This is very explicit while automatically removing utterances of the same text feels too "tricky" to me.

@zepumph what do you think? Is there anything else that should be done for this issue?

**Comment by zepumph** [comment]
I see that `delayTime` was converted in `alertStableDelay`. For example in ohms-law/ControlPanel. I wonder if this number is now too big, since the implementation should be improved now to keep these alerts in the queue more consistent. Mostly I can notice a delay between moving the slider and hearing the alert. Do you think those values need to be adjusted?

**Comment by zepumph** [comment]
- [ ] Should `alertStable` be renamed to `alertWhenStable`?


There seems to be an inherent flaw in the "alertStableDelay" which is that the "granularity" of that number is based on the step interval of the utteranceQueue. Since the step interval is `500`, `alertStableDelay` of 300 vs 400 won't have any effect. I wonder if we should separate the `stableTime` updating from the interval that sends items to the queue. Is this a worth while change at all? I'm trying to think of a situation where we would run into trouble because these values weren't being updated fast enough. Let's say that you have Utterance A (uA) and uB; uA has a delay of 400 and uB 300, uA is added to the utteranceQueue first and then uB. So you might think the ording will be uB and then uA because uB has a shorter wait time, but since it all resolves after 500 ms, UtteranceQueue.next() just picks the first item that fits the proper criteria (`if ( !alertStable || utteranceStabilized || alertMaximumDelay ) {`). This is a bit rambly but perhaps here are the two possible action steps:
- [ ] update `stableTime` more often than the utteranceQueue is stepped. 
- [ ] next() may want to support finding the"most appropriate" alert next, not just seeing if the most recent in the queue qualifies.

This is as far as I got, with this patch of a few changes for utterances and tests:

<details>

```
Index: js/accessibility/Utterance.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/accessibility/Utterance.js	(revision Local Version)
+++ js/accessibility/Utterance.js	(revision Shelved Version)
@@ -12,7 +12,7 @@
  * number of alerts associated with the utterance get read in order (see alert in options). Or it
  * may be that changes are being alerted rapidly from the same source. An Utterance is considered
  * "unstable" if it is being added rapidly to the utteranceQueue. By default, utterances are only
- * announced when they are "stable", and stop getting added to the queue. This will prevent 
+ * announced when they are "stable", and stop getting added to the queue. This will prevent
  * a large number of alerts from the same interaction from spamming the user. See related options
  * alertStable, alertStableDelay, and alertMaximumDelay.
  *
@@ -41,7 +41,7 @@
          * The content of the alert that this Utterance is wrapping. If it is an array, then the Utterance will
          * keep track of number of times that the Utterance has been alerted, and choose from the list "accordingly" see
          * loopingSchema for more details
-         * {AlertableDef}
+         * {string|Array.<string>}
          */
         alert: null,
 
@@ -94,7 +94,7 @@
       this.timeInQueue = 0;
 
       // @public (scenery-phet-internal) {number}  - in ms, how long this utterance has been "stable", which
-      // is the amount of time since this utterance has been added to the utteranceQueue
+      // is the amount of time since this utterance has been added to the utteranceQueue.
       this.stableTime = 0;
 
       // @public {boolean} - whether or not the utteranceQueue will wait alertStableDelay amount
@@ -103,7 +103,7 @@
       this.alertStable = options.alertStable;
 
       // @public (read-only, scenery-phet-internal) {number} - assign this utterance to a unique id so that
-      // we can suppress duplicates of this utterance in the utteranceQueue if alertStable is true
+      // we can suppress duplicates of this utterance in the utteranceQueue. Only supported if `alertStable` is true.
       this.uniqueId = instanceCount++;
 
       // @public {number} (scenery-phet-internal) - In ms, how long the utterance should remain in the queue before it
@@ -142,13 +142,17 @@
 
     /**
      * Set the alert for the utterance
-     * @param {AlertableDef} alert
+     * @param {string|Array.<string>} alert
      * @public
      */
     set alert( alert ) {
       this._alert = alert;
     }
 
+    get alert() {
+      return this._alert;
+    }
+
     /**
      * @public
      */
Index: js/accessibility/utteranceQueue.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/accessibility/utteranceQueue.js	(revision Local Version)
+++ js/accessibility/utteranceQueue.js	(revision Shelved Version)
@@ -40,7 +40,7 @@
       // initialized (cheers). See initialize();
       this._initialized = false;
 
-      // @private {Array.<Utterance>} - array of Utterances, spoken in first to last order
+      // @public (tests) {Array.<Utterance>} - array of Utterances, spoken in first to last order
       this.queue = [];
 
       // @private {number} the interval for sending alerts to the screen reader, in milliseconds
@@ -76,7 +76,7 @@
       }
 
       // clear any utterances if they are duplicates of the one being added
-      this.clearUtterances( utterance );
+      this.clearUtterance( utterance );
 
       this.queue.push( utterance );
     }
@@ -110,7 +110,7 @@
       }
 
       // remove any utterances if they are duplicates of the one being added
-      this.clearUtterances( utterance );
+      this.clearUtterance( utterance );
 
       this.queue.unshift( utterance );
     }
@@ -165,13 +165,13 @@
     }
 
     /**
-     * Called by addToFront and addToBack, do not call this. Clears the queue of all utterances duplicates
+     * Called by addToFront and addToBack, do not call this. Clears the queue of all duplicates of the provided Utterance
      * to support the behavior of alertStable. See Utterance.uniqueId for description of this feature.
      *
-     * @param {number} uniqueId
+     * @param {Utterance} utterance
      * @private
      */
-    clearUtterances( utterance ) {
+    clearUtterance( utterance ) {
 
       if ( utterance.alertStable ) {
 
@@ -256,6 +256,7 @@
     getStepInterval() {
       return this._stepInterval;
     }
+
     get stepInterval() { return this.getStepInterval(); }
 
     /**
@@ -263,7 +264,7 @@
      * @private
      */
     stepQueue() {
-
+      console.log( 'queue stepped' );
       // No-op function if the utteranceQueue is disabled
       if ( !this._enabled ) {
         return;
Index: js/accessibility/UtteranceTests.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/accessibility/UtteranceTests.js	(revision Local Version)
+++ js/accessibility/UtteranceTests.js	(revision Shelved Version)
@@ -41,7 +41,7 @@
 
       ariaHerald.initialize();
 
-      // whenever announcing, get a callback
+      // whenever announcing, get a callback and populate the alerts array
       ariaHerald.announcingEmitter.addListener( text => {
         alerts.unshift( text );
       } );
@@ -63,6 +63,35 @@
     }
   } );
 
+  QUnit.test( 'Utterance options', async assert => {
+
+    const myUtterance = new Utterance( {
+      alert: 'hi',
+      alertStable: true,
+      alertStableDelay: 1100
+    } );
+
+    for ( let i = 0; i < 100; i++ ) {
+
+      utteranceQueue.addToBack( myUtterance );
+    }
+
+    assert.ok( utteranceQueue.queue.length === 1, 'same Utterance should override in queue' );
+    await timeout( 600 );
+
+    assert.ok( myUtterance.stableTime >= utteranceQueue._stepInterval );
+
+    await timeout( 1000 );
+debugger;
+
+    assert.ok( utteranceQueue.queue.length === 0, 'same Utterance should override in queue' );
+    assert.ok( alerts.length === 1, 'same Utterance should override in queue' );
+    assert.ok( alerts[ 0 ] === myUtterance.alert, 'same Utterance should override in queue' );
+
+
+  } );
+
+
   QUnit.test( 'Basic Utterance testing', async assert => {
 
     // for this test, we just want to verify that the alert makes it through to ariaHerald
@@ -159,7 +188,7 @@
     const numAlerts = 4;
 
     // add the utterance to the back many times, by default they should collapse
-    for ( let i = 0; i < numAlerts; i++) {
+    for ( let i = 0; i < numAlerts; i++ ) {
       utteranceQueue.addToBack( highFrequencyUtterance );
     }
     assert.ok( utteranceQueue.queue.length === 1, 'utterances should collapse by default after addToBack' );

```

</details>

**Comment by jessegreenberg** [comment]
Good point, I hadn't thought of `alertStableDelay` as a way to order Utterances. Alternatively, now that `Utterance` manages timing for utterances maybe there isn't a need to have a large interval in utteranceQueue. Could that be reduced to something like ~10-20 ms so that `stableTime` can be updated with `utteranceQueue.next()`

**Comment by zepumph** [comment]
More review:

>alertMaximumDelay: Number.MAX_VALUE
- [ ] Is this the best default? I see alertStable is default to true, and the delat is default to 500 ms (quite a lot if you ask me). I think it is quite possible that the current default behavior of Utterance would yield nothing being alerted for a simple case like linking to a model Property that is changing every step. It could be nice to design the defaults geared more towards high frequency alerts. I'm on the fence about this though, because maybe `alertMaximumDelay` is more of a super use feature. If we feel that way, then I think we should decrease the default value of `alertStableDelay`. What do you think? Is there a reason you chose 500ms as a default for the delay? 

- [ ] addToBack and addToFront are 90% duplicated, could their implementations be factored out to the same function? This is separate but perhaps could be done with this work.


>      if ( nextUtterance && !this._muted && nextUtterance.predicate() ) {

- [ ] The last two of these feel like they should be their own function, like `this.canAlert` or something


- [ ] This may change, but right now timeInQueue and stableTime are basically the same thing, updated

**Comment by zepumph** [comment]
Back to you @jessegreenberg. I'm happy to go through some of these together, as some of these are more discussion points than recommendations. Let me know!

**Comment by jessegreenberg** [comment]
> I see that delayTime was converted in alertStableDelay...I wonder if this number is now too big

This was done in #491.

> There seems to be an inherent flaw in the "alertStableDelay" which is that the "granularity" of that number is based on the step interval of the utteranceQueue.

Also addressed in #491 as utteranceQueue is stepped every frame.

> alertMaximumDelay: Number.MAX_VALUE - Is this the best default?

I don't feel strongly about this. It was initially intended to be a "rarely used" feature, because I think Utterances describing Properties that change every animation frame will generally have more sim specific code for deciding when it is time to alert. But maybe ~3000 is a better default. Number.MAX_VALUE (or maybe Number.POSITIVE_INFINITY?) is a good default for ActivationUtterance (#492) so activation alerts are only heard once per activation.

> This may change, but right now timeInQueue and stableTime are basically the same thing, updated

Comment may not be finished, but the differences are important for the implementation of `alertMaximumDelay`. `timeInQueue` is the total time in queue. `stableTime` is the amount of time that has passed since the same Utterances was added to the queue.

> addToBack and addToFront are 90% duplicated,
> The last two of these feel like they should be their own function

Good points! Ill take a look.

**Comment by jessegreenberg** [comment]
Sorry it took so long to get to this. I did
> addToBack and addToFront are 90% duplicated,
The last two of these feel like they should be their own function

In the above two commits. I also checked off the review items that were discussed and handled in separate issues. @zepumph lets discuss the remaining check boxes on a11y dev meeting monday.

**Comment by jessegreenberg** [comment]
@zepumph and I paired on this today, and we went over the remaining review items. We identified that the remaining confusion about timeInQueue and stableTime is because  `next` was poorly named since it didn't always speak the "next" utterance. We distributed its work across new functions driven by `stepQueue` that are easier to understand.

We also identified that the option `alertStable` was not necessary, and decided that this feature will always  be in use when reusing Utterances for the queue. If ever you want to hear alerts as fast as possible (effectively alertStable: false) that can be done through an alertStableDelay of 0.

We also saw that  utterance.uniqueId was not necessary and it was replaced by checking against Utterance instances instead.
